akd/src/proto/mod.rs

```diff
@@ -0,0 +1,287 @@
+    fn from(input: &crate::helper_structs::Node<Hasher>) -> Self {
+        let mut result = Self::new();
+        result.set_label((&input.label).into());
+        result.set_hash(hash_to_bytes!(input.hash));
+        result
+    }
+}
+
+impl TryFrom<&audit::Node> for crate::helper_structs::Node<Hasher> {
+    type Error = String;
+
+    fn try_from(input: &audit::Node) -> Result<Self, Self::Error> {
+        require!(input, has_label);
+        require!(input, has_hash);
+        let label: crate::NodeLabel = input.get_label().try_into()?;
+        Ok(crate::helper_structs::Node::<Hasher> {
+            label,
+            hash: hash_from_bytes!(input.get_hash()),
+        })
+    }
+}
+
+impl From<&crate::proof_structs::SingleAppendOnlyProof<Hasher>> for audit::SingleEncodedProof {
+    fn from(input: &crate::proof_structs::SingleAppendOnlyProof<Hasher>) -> Self {
+        let mut result = Self::new();
+        let inserted = input
+            .inserted
+            .iter()
+            .map(|item| item.into())
+            .collect::<Vec<_>>();
+        let unchanged = input
+            .unchanged_nodes
+            .iter()
+            .map(|item| item.into())
+            .collect::<Vec<_>>();
+
+        result.set_inserted(RepeatedField::from_vec(inserted));
+        result.set_unchanged(RepeatedField::from_vec(unchanged));
+        result
+    }
+}
+
+impl TryFrom<audit::SingleEncodedProof> for crate::proof_structs::SingleAppendOnlyProof<Hasher> {
+    type Error = String;
+
+    fn try_from(input: audit::SingleEncodedProof) -> Result<Self, Self::Error> {
+        let mut inserted = vec![];
+        let mut unchanged = vec![];
+        for item in input.get_inserted() {
+            inserted.push(item.try_into()?);
+        }
+        for item in input.get_unchanged() {
+            unchanged.push(item.try_into()?);
+        }
+        Ok(crate::proof_structs::SingleAppendOnlyProof {
+            inserted,
+            unchanged_nodes: unchanged,
+        })
+    }
+}
+
+// ************************ Helper Functions ************************ //
+
+const NAME_SEPARATOR: char = '/';
+
+/// The constructed blobs with naming encoding the
+/// blob name = "EPOCH/PREVIOUS_ROOT_HASH/CURRENT_ROOT_HASH"
+pub struct AuditBlob {
+    /// The name of the blob, which can be decomposed into logical components (phash, chash, epoch)
+    pub name: String,
+    /// The binary data comprising the blob contents
+    pub data: Vec<u8>,
+}
+
+impl AuditBlob {
+    fn convert_akd_err(err: AkdError) -> String {
+        err.to_string()
+    }
+
+    /// Decompose the blob's name into the (previous_hash, current_hash, epoch) tuple
+    pub fn decompose_name(name: &str) -> Result<(Digest, Digest, u64), String> {
+        let parts = name.split(NAME_SEPARATOR).collect::<Vec<_>>();
+        if parts.len() < 3 {
+            return Err(String::from(
+                "Name is malformed, there are not enough components to reconstruct!",
+            ));
+        }
```

Comment:
With the new type comment in mind, it seems like we'd really clean this up (or maybe even remove this method entirely) depending on the implementation. Perhaps we can do something like this to be able to get to each of the parts within the type: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5fee50b5a9a6bb71eceb57f4c6f0798a

Of course, we can always just go with a struct here if we'd prefer names as well. Just a thought to clean up the implementation a bit and make this a little less error prone.